### ISSUE TITLE
Unable to get encoding from python plugin: failed to fill whole buffer

### ISSUE THREAD

**Issue by MatrixManAtYrService** [nushell/nushell#12141]
### Describe the bug

I'd like to write nushell plugins in python.  I was happy to see that there's an example in this repo, but I'm getting an error when I try to register it.

### How to reproduce

```
 ❯ nu --version
    0.90.1

 ❯ $env.PWD
    /home/matt/src/nushell/crates/nu_plugin_python
    
 ❯ git rev-parse HEAD
    430fb1fcb6f9c853ded1240e6ee8fb0e2e89669b

 ❯ register nu_plugin_python_example.py
    Error:   × Error getting signatures
       ╭─[entry #23:1:1]
     1 │ register nu_plugin_python_example.py
       · ────┬───
       ·     ╰── Plugin failed to load: unable to get encoding from plugin: failed to fill whole buffer
       ╰────
```

### Expected behavior

I expected the `register` command to complete, after which I could run `nu-python`.

### Configuration

| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.90.1                                                            |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | linux-x86_64                                                      |
| build_target       | x86_64-unknown-linux-gnu                                          |
| rust_version       | rustc 1.75.0 (82e1608df 2023-12-21) (built from a source tarball) |
| cargo_version      | cargo 1.75.0                                                      |
| build_time         | 1980-01-01 00:00:00 +00:00                                        |
| build_rust_channel | release                                                           |
| allocator          | mimalloc                                                          |
| features           | default, sqlite, trash, which, zip                                |
| installed_plugins  |                                                                   |

```
 ❯ cat /etc/os-release
BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
BUILD_ID="24.05.20231211.a9bf124"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 24.05 (Uakari)"
SUPPORT_URL="https://nixos.org/community.html"
VERSION="24.05 (Uakari)"
VERSION_CODENAME=uakari
VERSION_ID="24.05"
```

### Additional context

Nushell is awesome, thank you for maintaining it!

**Comment by fdncred** [comment]
@MatrixManAtYrService Devyn has made quite a few changes recently. Last I checked the python plugin was working.

@devyn any ideas?

**Comment by devyn** [comment]
@MatrixManAtYrService could you try on `0.91.0`? the python plugin was definitely broken in `0.90.x`

@fdncred ...but I think there may be another issue, which needs some maintainer attention - the version string in the python plugin will need to be updated with further nushell releases. I'm not sure if there's some workflow that does all of the version string updates for a release somewhere - if you can point me to it I'll fix it

**Comment by fdncred** [comment]
>  if you can point me to it I'll fix it

it's all just search and replace now. no workflow or action or anything like that.

**Comment by MatrixManAtYrService** [comment]
@devyn I've spent a little while trying different versions.  I'm not entirely sure how to interpret these results.  

The error message :point_up:  only happens when I have `0.90.1` (installed via nixpkgs) and I try to `register` the version of `nu_python_plugin_example.py` from `0.91.0`.  Using the wrong version was perhaps my mistake, although I'd hope for a nicer error.

When I use the corresponding version of `nu_python_plugin_example.py` (from `0.90.1`) the `register` command succeeds, but `nu-python` fails with:

```
Error:   × Error getting signatures
   ╭─[entry #14:1:1]
 1 │ register nu_plugin_python_example.py
   · ────┬───
   ·     ╰── Plugin failed to load: unable to get encoding from plugin: failed to fill whole buffer
   ╰────
```

This seems to be typical behavior for `0.90.1` (as seen :point_down:)
 
-----

To keep things in sync going forward, I switched go using `cargo run`.  I'm not rust savvy, so check me here, but here's what I did:
```
rm ~/.config/nushell/plugin.nu  # not sure if this is necessary
git checkout {the tag}
cargo run
cd crates/nu_plugin_python
register nu_plugin_python_example.py
nu-python 5 foo
```

And Here are my findings for several versions:

#### 0.90.1
`register` command works (unlike previously)
`nu-python` commands fails:

```
Error:   × Unable to decode call for nu-python
   ╭─[entry #3:1:1]
 1 │ nu-python 5 foo
   · ────┬────
   ·     ╰── Plugin failed to encode: missing field `val` at line 1 column 274
   ╰────
```

#### 0.91.0
`register` command fails:

```
Error:   × Error getting signatures
   ╭─[entry #1:1:1]
 1 │ register nu_plugin_python_example.py
   · ────┬───
   ·     ╰── Plugin failed to load: Plugin is compiled for nushell version 0.90.2, which is not compatible with version 0.91.0
   ╰────
```

#### main (430fb1fcb6f9c853ded1240e6ee8fb0e2e89669b)
`register` command works
`nu-python` command works

**Comment by fdncred** [comment]
You can't really mix and match versions of nushell and plugins. They need to be the same version.

**Comment by devyn** [comment]
@fdncred 

Yeah, but unfortunately it looks like the version string inside the python plugin didn't get updated for 0.91.0 - it's still set to 0.90.2. On `main` it is set to 0.91.1 as it should be, right now, so if the next release updates the version properly in that file too, it will be fine. The version string is now near the top of the file, in `NUSHELL_VERSION`. If you're using a glob or something to do it, make sure it's catching .py files too

@MatrixManAtYrService 

0.91.0 improved the error message for incompatible versions, so that should be a solved problem, but as I just mentioned I think it shipped with the python plugin reporting 0.90.2. If you change that in the python script to `"0.91.0"` (in `tell_nushell_hello()`), it should work properly

**Comment by MatrixManAtYrService** [comment]
Thanks @devyn, changing that string does indeed make it work with `0.91.0`.

This unblocks me so that I can play around with writing plugins, and it looks like going forward we can expect nicer errors too, so I think I'll close this.

I am curious about this @fdncred :

> You can't really mix and match versions of nushell and plugins. They need to be the same version.

Does this mean that if I upgrade nushell, my plugins will break?  Is there a recommended way of handling this from a plugin author's perspective? 

Naively I imagine that a plugin could support a range of versions by examining `$env.NU_VERSION` and providing different hello messages and different signatures based on what it sees, but maybe there's a better way?

**Comment by devyn** [comment]
@MatrixManAtYrService 

> > You can't really mix and match versions of nushell and plugins. They need to be the same version.
> 
> Does this mean that if I upgrade nushell, my plugins will break?  Is there a recommended way of handling this from a plugin author's perspective? 
> 
> Naively I imagine that a plugin could support a range of versions by examining `$env.NU_VERSION` and providing different hello messages and different signatures based on what it sees, but maybe there's a better way?

We expect that there might be breaking protocol changes between major versions, as we are allowed to make whatever breaking changes between major versions we want at the moment. We handle the comparison with semver and it's ideal if your plugin does too. Minor/patch version differences are considered to be compatible, and it doesn't matter whether the engine or plugin are higher.

When Nushell reaches 1.0, we will stabilize the plugin protocol and any new breaking changes will require an entry in the `features` key of `Hello` to be negotiated unless we release a new major version (i.e. 2.0.0)

You should report the version of Nushell you expect in `Hello` and do semver comparison on the one you receive.
